File: crates/ferrogate-cloudflare/src/d1.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-07-24
// description: Cloudflare D1 REST endpoint wrapper (database lifecycle + HTTP query API) for issue #420.

//! Cloudflare D1 REST endpoints (issue #420).
//!
//! A self-contained wrapper over the shared [`CloudflareClient`] (issue #405)
//! for the D1 database lifecycle and HTTP query endpoints:
//!
//! - `POST   /accounts/{account_id}/d1/database` — provision a database.
//! - `GET    /accounts/{account_id}/d1/database` — list databases.
//! - `GET    /accounts/{account_id}/d1/database/{uuid}` — fetch one database.
//! - `DELETE /accounts/{account_id}/d1/database/{uuid}` — delete a database.
//! - `POST   /accounts/{account_id}/d1/database/{uuid}/query` — run SQL.
//!
//! Auth, `{account_id}` templating, envelope decoding, typed error mapping,
//! and retry/backoff all come from the shared client, so this module is pure
//! endpoint shape: request/response DTOs plus thin methods.
//!
//! ## Rate-limit strategy (documented split, issue #420)
//!
//! The public REST API is limited to ~1,200 requests / 5 min / user, so this
//! raw-HTTP surface is the **admin / low-volume** path (provisioning, schema
//! migration, control-plane CRUD). Hot-path query traffic should go through a
//! **proxy Worker holding a D1 binding** (see
//! `docs/cloudflare-deploy-topology.md` §bindings and
//! `docs/cloudflare-integration.md` §6); that Worker is out of scope here.
//!
//! ## Parameter typing
//!
//! The documented REST schema types `params` as an array of **strings**.
//! SQLite column type affinity converts numeric-looking strings on insert
//! into `INTEGER`/`REAL` columns, and `NULL` is expressed in SQL (e.g.
//! `NULLIF(?, '')`) rather than as a JSON `null` param, so callers stay
//! inside the documented contract.

use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::client::{CloudflareClient, HttpMethod};
use crate::error::CloudflareError;

/// Request body for `POST /accounts/{account_id}/d1/database`.
#[derive(Debug, Clone, Serialize)]
pub struct D1CreateDatabaseRequest {
    /// D1 database name (unique per account).
    pub name: String,
    /// Placement hint (`wnam`/`enam`/`weur`/`eeur`/`apac`/`oc`). Ignored when
    /// `jurisdiction` is set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub primary_location_hint: Option<String>,
    /// Jurisdictional restriction (`eu`/`fedramp`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub jurisdiction: Option<String>,
}

impl D1CreateDatabaseRequest {
    /// A plain create request with no placement/jurisdiction constraints.
    pub fn named(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            primary_location_hint: None,
            jurisdiction: None,
        }
    }
}

/// A D1 database descriptor as returned by the lifecycle endpoints. All
/// fields are optional in Cloudflare's schema; `uuid` is the identifier the
/// query endpoint routes on.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct D1Database {
    #[serde(default)]
    pub uuid: Option<String>,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub version: Option<String>,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub file_size: Option<u64>,
    #[serde(default)]
    pub num_tables: Option<u64>,
}

/// The `meta` object attached to each query result: row accounting and the
/// write cursor (`last_row_id`), all optional in Cloudflare's schema.
#[derive(Debug, Clone, Default, Deserialize, PartialEq)]
pub struct D1QueryMeta {
    #[serde(default)]
    pub changed_db: Option<bool>,
    #[serde(default)]
    pub changes: Option<u64>,
    #[serde(default)]
    pub duration: Option<f64>,
    #[serde(default)]
    pub last_row_id: Option<i64>,
    #[serde(default)]
    pub rows_read: Option<u64>,
    #[serde(default)]
    pub rows_written: Option<u64>,
    #[serde(default)]
    pub size_after: Option<u64>,
    #[serde(default)]
    pub served_by_region: Option<String>,
}

/// One statement's outcome from the query endpoint. `results` holds the rows
/// as JSON objects keyed by column name; a multi-statement batch yields one
/// [`D1QueryResult`] per statement.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct D1QueryResult {
    #[serde(default)]
    pub results: Vec<serde_json::Value>,
    #[serde(default)]
    pub success: Option<bool>,
    #[serde(default)]
    pub meta: Option<D1QueryMeta>,
}

impl D1QueryResult {
    /// `meta.changes` collapsed to a plain count (0 when absent).
    pub fn changes(&self) -> u64 {
        self.meta
            .as_ref()
            .and_then(|meta| meta.changes)
            .unwrap_or_default()
    }
}

#[derive(Serialize)]
struct D1QueryRequest<'a> {
    sql: &'a str,
    params: &'a [String],
}

/// Thin D1 endpoint client over the shared [`CloudflareClient`].
#[derive(Clone)]
pub struct D1Client {
    client: Arc<CloudflareClient>,
}

impl D1Client {
    pub fn new(client: Arc<CloudflareClient>) -> Self {
        Self { client }
    }

    /// The shared client this wrapper routes through.
    pub fn client(&self) -> &Arc<CloudflareClient> {
        &self.client
    }

    /// Provision a new D1 database. Returns the descriptor whose `uuid` the
    /// query endpoint routes on.
    pub async fn create_database(
        &self,
        request: &D1CreateDatabaseRequest,
    ) -> Result<D1Database, CloudflareError> {
        let body = serde_json::to_vec(request).map_err(|error| {
            CloudflareError::Config(format!("failed to encode D1 create request: {error}"))
        })?;
        self.client
            .request_json(
                HttpMethod::Post,
                "accounts/{account_id}/d1/database",
                Some(body),
                None,
            )
            .await
    }

    /// List the account's D1 databases. Requests the maximum page size
    /// (1,000); pagination beyond the first page is intentionally not
    /// followed in this admin-path slice.
    pub async fn list_databases(&self) -> Result<Vec<D1Database>, CloudflareError> {
        self.client
            .get_json("accounts/{account_id}/d1/database?per_page=1000", None)
            .await
    }

    /// Fetch one database descriptor by uuid.
    pub async fn get_database(&self, database_id: &str) -> Result<D1Database, CloudflareError> {
        let path = database_path(database_id, "")?;
        self.client.get_json(&path, None).await
    }

    /// Delete a database (and all of its data). The endpoint returns a null
    /// `result`, so this is an ack-style call.
    pub async fn delete_database(&self, database_id: &str) -> Result<(), CloudflareError> {
        let path = database_path(database_id, "")?;
        self.client
            .request_ack(HttpMethod::Delete, &path, None, None)
            .await
    }

    /// Run SQL against a database over the HTTP query endpoint.
    ///
    /// `sql` may contain multiple statements joined by semicolons (executed
    /// as a batch, one [`D1QueryResult`] per statement). `params` bind `?`
    /// placeholders positionally and are strings per the documented schema
    /// (see the module docs on affinity/NULL handling). Limits: statement
    /// <= 100 KB, <= 100 params, 30 s runtime, single-threaded per database.
    pub async fn query(
        &self,
        database_id: &str,
        sql: &str,
        params: &[String],
    ) -> Result<Vec<D1QueryResult>, CloudflareError> {
        let path = database_path(database_id, "/query")?;
        let body = serde_json::to_vec(&D1QueryRequest { sql, params }).map_err(|error| {
            CloudflareError::Config(format!("failed to encode D1 query request: {error}"))
        })?;
        self.client
            .request_json(HttpMethod::Post, &path, Some(body), None)
            .await
    }
}

/// Build the `d1/database/{uuid}{suffix}` path, rejecting ids that could
/// escape the path segment. Cloudflare ids are hex uuids; anything else is a
/// caller bug surfaced as a config error before any request is sent.
fn database_path(database_id: &str, suffix: &str) -> Result<String, CloudflareError> {
    if database_id.is_empty()
        || !database_id
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '-')
    {
        return Err(CloudflareError::Config(format!(
            "invalid D1 database id {database_id:?}: expected a Cloudflare uuid"
        )));
    }
    Ok(format!(
        "accounts/{{account_id}}/d1/database/{database_id}{suffix}"
    ))
}

File: packages/cloudflare/src/d1.ts
/**
 * D1 database **LIFECYCLE** — slice **S5**.
 *
 * Ported from `crates/ferrogate-cloudflare/src/d1.rs:159-219`.
 *
 * ## Why this exists when "D1 is a binding"
 *
 * `cutover-parity-libraries.md` §6.1 classified ALL of `d1.rs` as "superseded by
 * the native D1 binding" and therefore never gave it a port-plan row. That is
 * true of the `/query` endpoint and **false of the lifecycle endpoints**: no
 * binding can CREATE a D1 database, for the same reason no binding can create
 * an R2 bucket. `env.DB` addresses a database that already exists.
 *
 * FerroGate's design is **one D1 database per tenant**, and nothing in the TS
 * tree creates one: `EnvBindingTenantDatabaseRouter` resolves
 * `tenantId → binding` through the control DB's `tenant_databases` table and
 * fails closed on a miss, but every reference to that table is a READER. The
 * documented onboarding procedure is a comment block instructing an operator to
 * run `wrangler d1 create`, hand-write an `INSERT INTO tenant_databases`, add a
 * `[[d1_databases]]` stanza and redeploy. This module removes the manual half.
 *
 * ## The `/query` endpoint is deliberately NOT here
 *
 * `@ferrogate/storage`'s `D1RestDatabase` already implements it, is exercised
 * against a real `workerd`, and models the atomicity limits the REST path has
 * (no transaction envelope → `batch()` refuses). A second copy here would be
 * exactly the duplication this package exists to end. Provisioning-time
 * migrations should run through that class.
 *
 * ## The deploy-time constraint this does NOT remove
 *
 * Creating a database is programmable; **binding** it is not. A newly created
 * database still needs a `[[d1_databases]]` stanza and a deploy before
 * `EnvBindingTenantDatabaseRouter` can route to it. That is the standing open
 * constraint on the whole one-DB-per-tenant design. The deploy-free alternative
 * is `NonAtomicD1RestTenantDatabaseRouter`, which addresses a database by
 * runtime uuid — and refuses the money paths for it.
 *
 * ## Wiring line
 *
 * From a control-plane tenant-onboarding handler (never the request path):
 *
 * ```ts
 * const database = await new D1LifecycleClient(cf).createDatabase({ name });
 * // then: run sql/d1-ts/tenant/* through @ferrogate/storage's D1RestDatabase,
 * // then INSERT the tenant_databases row, then add the binding + deploy.
 * ```
 */
import type { CloudflareClient } from "./client.js";
import { CloudflareError } from "./errors.js";

const D1_DATABASES_PATH = "accounts/{account_id}/d1/database";

/**
 * The list endpoint caps `per_page` at 1,000. D1's list is PAGE-NUMBERED
 * (`page`/`per_page`), unlike R2's cursor dialect; confusing the two silently
 * returns page 1.
 */
const D1_PER_PAGE = 1000;

/** Request body for `POST /accounts/{account_id}/d1/database`. */
export interface D1CreateDatabaseRequest {
  /** D1 database name, unique per account. */
  readonly name: string;
  /** `wnam`/`enam`/`weur`/`eeur`/`apac`/`oc`. Ignored when `jurisdiction` is set. */
  readonly primaryLocationHint?: string;
  /** Jurisdictional restriction (`eu`/`fedramp`). */
  readonly jurisdiction?: string;
}

/** A D1 database descriptor. All fields are optional in Cloudflare's schema. */
export interface D1DatabaseDescriptor {
  /** The identifier the query endpoint routes on — `tenant_databases.database_uuid`. */
  readonly uuid?: string;
  readonly name?: string;
  readonly version?: string;
  readonly created_at?: string;
  readonly file_size?: number;
  readonly num_tables?: number;
}

/** The D1 database-lifecycle surface over the shared client. */
export class D1LifecycleClient {
  constructor(private readonly client: CloudflareClient) {}

  /**
   * Provision a new D1 database. Returns the descriptor whose `uuid` the query
   * endpoint routes on and the `tenant_databases` registry row records.
   *
   * Retry is opted IN: D1 database names are unique per account, so a re-issued
   * create after a 5xx cannot silently produce a SECOND database — it either
   * succeeds or reports the duplicate. That is the property the token mint in
   * `r2-token.ts` lacks, which is why that one opts out.
   */
  async createDatabase(request: D1CreateDatabaseRequest): Promise<D1DatabaseDescriptor> {
    if (request.name.trim() === "") {
      throw CloudflareError.config("D1 create-database requires a non-empty name");
    }
    const body: Record<string, string> = { name: request.name };
    if (request.primaryLocationHint !== undefined) {
      body.primary_location_hint = request.primaryLocationHint;
    }
    if (request.jurisdiction !== undefined) body.jurisdiction = request.jurisdiction;
    return this.client.requestJson<D1DatabaseDescriptor>("POST", D1_DATABASES_PATH, {
      body,
      idempotent: true,
    });
  }

  /**
   * List **all** of the account's D1 databases, walking pages beyond the first.
   *
   * Terminates once a page returns fewer rows than `per_page` — no cursor state.
   * Costs one extra empty page only in the exact-multiple case.
   */
  async listDatabases(): Promise<D1DatabaseDescriptor[]> {
    const databases: D1DatabaseDescriptor[] = [];
    let page = 1;
    for (;;) {
      const batch = await this.client.getJson<D1DatabaseDescriptor[]>(
        `${D1_DATABASES_PATH}?per_page=${D1_PER_PAGE}&page=${page}`,
      );
      databases.push(...batch);
      if (batch.length < D1_PER_PAGE) return databases;
      page += 1;
    }
  }

  /** Fetch one database descriptor by uuid. */
  async getDatabase(databaseId: string): Promise<D1DatabaseDescriptor> {
    return this.client.getJson<D1DatabaseDescriptor>(databasePath(databaseId));
  }

  /** Delete a database **and all of its data**. */
  async deleteDatabase(databaseId: string): Promise<void> {
    await this.client.requestAck("DELETE", databasePath(databaseId), { idempotent: true });
  }
}

/**
 * Build the `d1/database/{uuid}` path, rejecting ids that could escape the path
 * segment. Cloudflare ids are hex uuids; anything else is a caller bug surfaced
 * before any request is sent.
 */
function databasePath(databaseId: string): string {
  if (databaseId === "" || !/^[a-zA-Z0-9-]+$/.test(databaseId)) {
    throw CloudflareError.config(
      `invalid D1 database id ${JSON.stringify(databaseId)}: expected a Cloudflare uuid`,
    );
  }
  return `${D1_DATABASES_PATH}/${databaseId}`;
}
